File: backend/app/schemas/stock.py
"""Pydantic schemas for stock data."""
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class StockQuote(BaseModel):
    ticker: str
    name: str
    price: float
    change: float
    change_pct: float
    volume: int
    avg_volume: int
    market_cap: Optional[float] = None
    pe_ratio: Optional[float] = None
    fifty_two_week_high: float
    fifty_two_week_low: float
    timestamp: datetime

    model_config = {"from_attributes": True}


class OHLCVBar(BaseModel):
    timestamp: datetime
    open: float
    high: float
    low: float
    close: float
    volume: int


class HistoricalData(BaseModel):
    ticker: str
    range: str
    interval: str
    data: List[OHLCVBar]


class Fundamentals(BaseModel):
    ticker: str
    pe_ratio: Optional[float] = None
    eps: Optional[float] = None
    roe: Optional[float] = None
    debt_to_equity: Optional[float] = None
    free_cash_flow: Optional[float] = None
    beta: Optional[float] = None
    gross_margin: Optional[float] = None
    operating_margin: Optional[float] = None

File: frontend/src/types/stock.ts
// Stock types
export interface StockQuote {
  ticker: string;
  name: string;
  price: number;
  change: number;
  change_pct: number;
  volume: number;
  avg_volume: number;
  market_cap: number | null;
  pe_ratio: number | null;
  fifty_two_week_high: number;
  fifty_two_week_low: number;
  timestamp: string;
}

export interface OHLCVBar {
  timestamp: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface HistoricalData {
  ticker: string;
  range: string;
  interval: string;
  data: OHLCVBar[];
}

export interface Fundamentals {
  ticker: string;
  sector: string | null;
  industry: string | null;
  pe_ratio: number | null;
  forward_pe: number | null;
  eps: number | null;
  revenue: number | null;
  gross_margin: number | null;
  operating_margin: number | null;
  roe: number | null;
  debt_to_equity: number | null;
  free_cash_flow: number | null;
  dividend_yield: number | null;
  beta: number | null;
}

export interface NewsItem {
  title: string;
  source: string;
  url: string;
  published_at: string;
  sentiment_score: number;
  sentiment_label: 'positive' | 'neutral' | 'negative';
}
